programs/Makefile

```diff
@@ -92,12 +84,19 @@ ifneq (,$(filter Windows%,$(OS)))
 endif
 	$(CC) $(FLAGS) $^ $(RES_FILE) -o zstd$(EXT) $(LDFLAGS)
 
+zstd-nogz : CPPFLAGS += -DZSTD_LEGACY_SUPPORT=$(ZSTD_LEGACY_SUPPORT)
 zstd-nogz : HAVE_ZLIB=0
 zstd-nogz : zstd-internal
 
-zstd : CPPFLAGS += $(ZLIBCPP)
-zstd : LDFLAGS += $(ZLIBLD)
-zstd : zstd-internal
+zstd : CPPFLAGS += -DZSTD_LEGACY_SUPPORT=$(ZSTD_LEGACY_SUPPORT)
+zstd :
+	@echo "int main(){}" | $(CC) -o have_zlib -x c - -lz && echo zlib found || echo zlib not found
+	@if [ -s have_zlib ]; then \
+         $(RM) have_zlib$(EXT) fileio.o \
+         && $(MAKE) zstd-internal HAVE_ZLIB=1 CFLAGS="$(CFLAGS)" CPPFLAGS="$(CPPFLAGS) -DZSTD_GZDECOMPRESS " LDFLAGS="$(LDFLAGS) -lz" ; \
```

Comment:
This solution calls `make` inside `make`. While it works, I prefer to avoid it when possible.

I think you managed to control the implied limitation (second `make` cannot modify environment variables), so I believe it does work correctly.

The issue is more about how evolutive the construction is, notably with nested targets, considering here the possibility of additional cli variants.